contracts/root/StakeManager.sol

```diff
@@ -6,182 +6,244 @@ import { SafeMath } from "openzeppelin-solidity/contracts/math/SafeMath.sol";
+import { ValidatorSet } from "./ValidatorSet.sol"; 
 
 
 contract StakeManager is StakeManagerInterface, RootChainable, Lockable {
   using SafeMath for uint256;
+  using SafeMath for uint8;
   using ECVerify for bytes32;
 
+  PriorityQueue stakeQueue;
+  ValidatorSet validatorSet;
   // token object
   ERC20 public tokenObj;
 
-  // validator threshold
-  uint256 public _validatorThreshold = 0;
-
-  // pool of stakers and validators
-  struct Stake {
-    uint256 amount;
-    address staker;
-  }
-
-  // map of stakers (address => staked index)
-  mapping(address => uint256) public stakers;
+  // The randomness seed of the epoch.
+  // This is used to determine the proposer and the validator pool
+  bytes32 public epochSeed = keccak256(abi.encodePacked(block.difficulty + block.number + now));
 
-  // stake list
-  Stake[] public stakeList;
+  event ThresholdChange(uint256 newThreshold, uint256 oldThreshold);
+  //optional event to ack unstaking
+  event UnstakeInit(address indexed user, uint256 amount, uint256 total, bytes data); 
 
+  // validator threshold
+  uint256 public _validatorThreshold = 0;
   // total stake
-  uint256 public totalStake;
+  uint256 public totalStake = 0;
+  // current epoch
+  uint256 public currentEpoch = 0;
 
-  // The randomness seed of the epoch.
-  // This is used to determine the proposer and the validator pool
-  bytes32 public epochSeed = keccak256(abi.encodePacked(block.difficulty + block.number + now));
+  //Todo: dynamically update
+  uint256 public minStakeAmount = 0;
+  uint256 public minLockInPeriod = 1; //(unit epochs)
+  uint256 public stakingIdCount = 0;  // just a counter/index to map it with PQ w/address
 
-  //
-  // Events
-  //
+  enum State { isStaker, isValidator, isProposer, isBaned, isUnstaking }
 
-  event ThresholdChange(uint256 newThreshold, uint256 oldThreshold);
+  struct Staker {
+    uint256 epoch;  // init 0 
+    uint256 amount;
+    bytes data;
+    State state;
+    uint256 stakingId; 
+  }
+
+  struct StakeExit {
+    uint256 amount;
+    address stakerAddress;
+  }
 
-  // ERC900
-  event Staked(address indexed user, uint256 amount, uint256 total, bytes data);
-  event Unstaked(address indexed user, uint256 amount, uint256 total, bytes data);
+  StakeExit[] exiterList;
+  address[] stakersList;
+  address[] public currentValidators;
 
-  //
-  // Constructor
-  //
+  mapping (address=>Staker) stakers; 
+  mapping (uint256=>address) stakingIdToAddress;
 
-  constructor (address _token) public {
+  constructor(address _token) public payable {
     require(_token != 0x0);
     tokenObj = ERC20(_token);
+    stakeQueue = new PriorityQueue();
   }
 
-  //
-  // Modifiers
-  //
-
   // only staker
   modifier onlyStaker() {
     require(totalStakedFor(msg.sender) > 0);
     _;
   }
 
-  //
-  // ERC900 implementation
-  //
+  
+  function _priority(address user, uint256 amount, bytes data)internal view returns(uint256){
+    // priority = priority << 64 | amount.div(totalStake) ;
```

Comment:
Please explain comments

Reply:
Shouldn't function be `pure` instead of view?

Reply:
priority calculation logic is commented, need to test it's correctness. After that will update priority function.